### ISSUE TITLE
Multi-component systems lose charges in subsequent components

### ISSUE THREAD

**Issue by davidlmobley** [open-forcefield-group/smarty#186]
I'm working on #184, prototyping conversion of the cyclohexane/ethanol mixed system which is provided here into AMBER via ParmEd, and uncovered a bug where subsequent molecules (after the first one) in a system end up with zero charges. I'll troubleshoot.

**Comment by jchodera** [comment]
You'll want to focus on [these lines](https://github.com/open-forcefield-group/smarty/blob/master/smarty/forcefield.py#L1858-L1864) and add some debug code to see if it isn't finding all the matches or if somehow the charges in the `OEMol` are getting wiped out.

**Comment by davidlmobley** [comment]
I've basically isolated the problem, which I think is on this line:
`for (atom, atom_index) in zip(reference_molecule.GetAtoms(), atom_mapping):`

This pulls a specific key from the `atom_mappings` dictionary, but the key is never used to loop over the various mappings.

I should be able to fix this easily, but will be offline for the evening so it will have to wait til the morning.

**Comment by jchodera** [comment]
`atom_mappings` should be a list of atom indices, not a dictionary, [according to the documentation](https://github.com/open-forcefield-group/smarty/blob/master/smarty/forcefield.py#L197).

**Comment by davidlmobley** [comment]
Well,  I didn't look at the docs; I printed the contents and it was a dictionary. And it needs to be unrolled more than it is being right now.

**Comment by jchodera** [comment]
Yep, you're right. This is a case of having confused myself by not keeping the documentation consistent with the code. Suggest we:
* Change 
```python
for (atom, atom_index) in zip(reference_molecule.GetAtoms(), atom_mapping):
```
to
```python
for (atom, atom_index) in atom_mapping.items():
```
which should give the correct behavior.
* Update the documentation on [this line](https://github.com/open-forcefield-group/smarty/blob/master/smarty/forcefield.py#L197). It should be a `list` of `dict`s, where each `dict` maps the `OEAtom` onto a topology atom index.
* Add a test based on the system you're currently working with to check if some known molecules receive the same charges.

**Comment by davidlmobley** [comment]
@jchodera - not quite, as the code there assumes that `atom` is an OEAtom, whereas in the dictionary it is an index. That is, if I print the dictionary contents they look something like:
```
'Atom mappings:', [{0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10, 11: 11, 12: 12, 13: 13, 14: 14, 15: 15, 16: 16, 17: 17}, {0: 18, 1: 19, 2: 20, 3: 21, 4: 22, 5: 23, 6: 24, 7: 25, 8: 26, 9: 27, 10: 28, 11: 29, 12: 30, 13: 31, 14: 32, 15: 33, 16: 34, 17: 35} .... ]
```
Troubleshooting.

**Comment by jchodera** [comment]
Ah, OK. You'd need to use
```python
atoms = [ atom for atom in reference_molecule.GetAtoms() ]
for (atom, atom_index) in atom_mapping.items():
```

**Comment by davidlmobley** [comment]
I have a fix, though it's not that. The issues were deeper.

(Note that your code snippet does not use the `atoms` list that you just defined in the `for` loop, so it doesn't help. I've had to add code to retrieve the atoms from the reference molecule, store their charges, and then look them up by the atom index. I'll push an update to my PR soon after I do some cleanup.)

**Comment by davidlmobley** [comment]
I have a fix for this in #187 now, though I don't have a test built in yet. (For a test, I'll validate energies against AMBER for a system with multiple components, but I need to get something else working before I get to that.)
